cleverhans/attacks_tf.py

```diff
@@ -257,151 +184,138 @@ def jacobian_graph(predictions, x, nb_classes):
-                                        adv_x_original_shape, feed=feed)
+        target_class = tf.reshape(tf.transpose(y_in, perm=[1, 0]),
+                                  shape=[nb_classes, -1, 1])
+        other_classes = tf.cast(tf.not_equal(target_class, 1), tf.float32)
+
+        grads_target = tf.reduce_sum(grads * target_class, axis=0)
+        grads_other = tf.reduce_sum(grads * other_classes, axis=0)
+
+        # Remove the already-used input features from the search space
+        # Subtract 2 times the maximum value from those value so that
+        # they won't be picked later
+        increase_coef = (4 * int(increase) - 2) \
+            * tf.cast(tf.equal(domain_in, 0), tf.float32)
+
+        target_tmp = grads_target
+        target_tmp -= increase_coef \
+            * tf.reduce_max(tf.abs(grads_target), axis=1, keep_dims=True)
+        target_sum = tf.reshape(target_tmp, shape=[-1, nb_features, 1]) \
+            + tf.reshape(target_tmp, shape=[-1, 1, nb_features])
+
+        other_tmp = grads_other
+        other_tmp += increase_coef \
+            * tf.reduce_max(tf.abs(grads_other), axis=1, keep_dims=True)
+        other_sum = tf.reshape(other_tmp, shape=[-1, nb_features, 1]) \
+            + tf.reshape(other_tmp, shape=[-1, 1, nb_features])
+
+        # Create a mask to only keep features that match conditions
+        if increase:
+            scores_mask = ((target_sum > 0) & (other_sum < 0))
+        else:
+            scores_mask = ((target_sum < 0) & (other_sum > 0))
 
-        # Update loop variables
-        iteration = iteration + 1
+        # Create a 2D numpy array of scores for each pair of candidate features
+        scores = tf.cast(scores_mask, tf.float32) \
+            * (-target_sum * other_sum) * zero_diagonal
 
-    if current == target:
-        _logger.info("Attack succeeded using {} iterations".format(iteration))
-    else:
-        _logger.info(("Failed to find adversarial example " +
-                      "after {} iterations").format(iteration))
+        # Extract the best two pixels
+        best = tf.argmax(
+                    tf.reshape(scores, shape=[-1, nb_features * nb_features]),
+                    axis=1)
 
-    # Compute the ratio of pixels perturbed by the algorithm
-    percent_perturbed = float(iteration * 2) / nb_features
+        p1 = tf.mod(best, nb_features)
+        p2 = tf.floordiv(best, nb_features)
+        p1_one_hot = tf.one_hot(p1, depth=nb_features)
+        p2_one_hot = tf.one_hot(p2, depth=nb_features)
 
-    # Report success when the adversarial example is misclassified in the
-    # target class
-    if current == target:
-        return np.reshape(adv_x, original_shape), 1, percent_perturbed
-    else:
-        return np.reshape(adv_x, original_shape), 0, percent_perturbed
+        # Check if more modification is needed for each samples
+        mod_done = tf.equal(tf.reduce_sum(y_in * preds_onehot, axis=1), 0)
```

Comment:
Should this be called `mod_not_done` instead?

Reply:
Good catch. Will fix.